packages/database-mongo-magic-link/src/mongo-magic-link.ts

```diff
@@ -0,0 +1,154 @@
+import { Collection, Db, IndexOptions } from 'mongodb';
+import { DatabaseInterfaceServiceMagicLink, User } from '@accounts/types';
+import { toMongoID } from './utils';
+import { MongoServicePassword, MongoServicePasswordOptions } from '@accounts/mongo-password';
+
+export interface MongoUser {
+  _id?: string | object;
+  username?: string;
+  services: {
+    token?: {
+      token: string;
+    };
+  };
+  emails?: [
+    {
+      address: string;
+      verified: boolean;
+    }
+  ];
+
+  [key: string]: any;
+}
+
+export interface MongoServiceMagicLinkOptions {
+  /**
+   * Mongo database object.
+   */
+  database: Db;
+  /**
+   * The users collection name.
+   * Default 'users'.
+   */
+  userCollectionName?: string;
+  /**
+   * The timestamps for the users collection.
+   * Default 'createdAt' and 'updatedAt'.
+   */
+  timestamps?: {
+    createdAt: string;
+    updatedAt: string;
+  };
+  /**
+   * Should the user collection use _id as string or ObjectId.
+   * Default 'true'.
+   */
+  convertUserIdToMongoObjectId?: boolean;
+  /**
+   * Function that generate the id for new objects.
+   */
+  idProvider?: () => string | object;
+  /**
+   * Function that generate the date for the timestamps.
+   * Default to `(date?: Date) => (date ? date.getTime() : Date.now())`.
+   */
+  dateProvider?: (date?: Date) => any;
+
+  /**
+   * Mongo password service
+   */
+  password: MongoServicePassword;
+}
+
+const defaultOptions = {
+  userCollectionName: 'users',
+  dateProvider: (date?: Date) => (date ? date.getTime() : Date.now()),
+};
+
+export class MongoServiceMagicLink implements DatabaseInterfaceServiceMagicLink {
+  // Merged options that can be used
+  private options: MongoServiceMagicLinkOptions & typeof defaultOptions;
+  // Mongo database object
+  private database: Db;
+  // Mongo user collection
+  private userCollection: Collection;
+  // Password service
+  private password: MongoServicePassword;
+
+  constructor(options: MongoServiceMagicLinkOptions) {
+    const passwordOptions = (options.password as any).options as MongoServicePasswordOptions;
+    this.options = {
+      ...defaultOptions,
+      ...passwordOptions,
+      ...options,
+    };
+
+    this.database = this.options.database;
+    this.userCollection = this.database.collection(this.options.userCollectionName);
+    this.password = this.options.password;
+  }
+
+  /**
+   * Setup the mongo indexes needed for the token service.
+   * @param options Options passed to the mongo native `createIndex` method.
+   */
+  public async setupIndexes(options: Omit<IndexOptions, 'unique' | 'sparse'> = {}): Promise<void> {
+    // Token index used to verify the email address of a user
+    await this.userCollection.createIndex('services.token.loginTokens.token', {
```

Comment:
should we change token to magicLik here?

Reply:
The service name should definately change to magicLink - was that what you meant?